// @pm2/io/build/main/transports/IPCTransport.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var Debug = require("debug");
var eventemitter2_1 = require("eventemitter2");
var cluster = require("cluster");
var EventLoopInspector = require("event-loop-inspector");
var IPCTransport = (function (_super) {
    tslib_1.__extends(IPCTransport, _super);
    function IPCTransport() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.initiated = false;
        _this.logger = Debug('axm:transport:ipc');
        _this.eventLoopInspector = EventLoopInspector(false);
        return _this;
    }
    IPCTransport.prototype.init = function (config) {
        var _this = this;
        this.logger('Init new transport service');
        if (this.initiated === true) {
            console.error("Trying to re-init the transport, please avoid");
            return this;
        }
        this.initiated = true;
        this.logger('Agent launched');
        this.onMessage = function (data) {
            _this.logger("Received reverse message from IPC");
            _this.emit('data', data);
        };
        process.on('message', this.onMessage);
        if (cluster.isWorker === false) {
            this.autoExitHook();
        }
        return this;
    };
    IPCTransport.prototype.autoExitHook = function () {
        var _this = this;
        this.autoExitHandle = setInterval(function () {
            var dump = _this.eventLoopInspector.dump();
            var requests = Object.keys(dump.requests);
            var handles = Object.keys(dump.handles);
            var sockets = dump.handles.Socket;
            var pipes = dump.handles.Pipe;
            var hasStdOut = sockets && sockets.find(function (sock) { return sock.fd === 1; }) !== undefined;
            var hasStdErr = sockets && sockets.find(function (sock) { return sock.fd === 2; }) !== undefined;
            var isSocketStds = sockets && sockets.length === 2 && hasStdErr && hasStdOut;
            var isOnlySocketStds = handles.length === 1 && isSocketStds;
            var isStdsWithIPC = isSocketStds && pipes && pipes.length === 1 && handles.length === 2;
            var shouldStillExit = isStdsWithIPC || isOnlySocketStds;
            if ((handles.length > 0 || requests.length > 0) && shouldStillExit === false) {
                return _this.logger('no need to exit, there are handles/requests in uv', JSON.stringify(dump));
            }
            _this.logger("Nothing found in uv, removing the IPC listener");
            process.removeListener('message', _this.onMessage);
            var tmp = setTimeout(function (_) {
                _this.logger("Still alive, listen back to IPC");
                process.on('message', _this.onMessage);
            }, 200);
            tmp.unref();
        }, 5000);
        this.autoExitHandle.unref();
    };
    IPCTransport.prototype.setMetrics = function (metrics) {
        var serializedMetric = metrics.reduce(function (object, metric) {
            if (typeof metric.name !== 'string')
                return object;
            object[metric.name] = {
                historic: metric.historic,
                unit: metric.unit,
                type: metric.id,
                value: metric.value
            };
            return object;
        }, {});
        this.send('axm:monitor', serializedMetric);
    };
    IPCTransport.prototype.addAction = function (action) {
        this.logger("Add action: " + action.name + ":" + action.type);
        this.send('axm:action', {
            action_name: action.name,
            action_type: action.type,
            arity: action.arity,
            opts: action.opts
        });
    };
    IPCTransport.prototype.setOptions = function (options) {
        this.logger("Set options: [" + Object.keys(options).join(',') + "]");
        return this.send('axm:option:configuration', options);
    };
    IPCTransport.prototype.send = function (channel, payload) {
        if (typeof process.send !== 'function')
            return -1;
        try {
            process.send({ type: channel, data: payload });
        }
        catch (err) {
            this.logger('Process disconnected from parent !');
            this.logger(err);
            return process.exit(1);
        }
    };
    IPCTransport.prototype.destroy = function () {
        if (this.onMessage !== undefined) {
            process.removeListener('message', this.onMessage);
        }
        if (this.autoExitHandle !== undefined) {
            clearInterval(this.autoExitHandle);
        }
        this.logger('destroy');
    };
    return IPCTransport;
}(eventemitter2_1.EventEmitter2));
exports.IPCTransport = IPCTransport;
