// @pm2/io/build/main/transports/WebsocketTransport.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var semver = require("semver");
var debug_1 = require("debug");
var eventemitter2_1 = require("eventemitter2");
var transactionAggregator_1 = require("../utils/transactionAggregator");
var SerializedAction = (function () {
    function SerializedAction() {
    }
    return SerializedAction;
}());
var ProcessMetadata = (function () {
    function ProcessMetadata() {
    }
    return ProcessMetadata;
}());
exports.ProcessMetadata = ProcessMetadata;
var WebsocketTransport = (function (_super) {
    tslib_1.__extends(WebsocketTransport, _super);
    function WebsocketTransport() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.initiated = false;
        _this.logger = debug_1.default('axm:transport:websocket');
        return _this;
    }
    WebsocketTransport.prototype.init = function (config) {
        var _this = this;
        if (!semver.satisfies(process.version, '>= 6.0.0')) {
            console.error('[STANDALONE MODE] Unable to set standalone mode with node < 6.0.0');
            return process.exit(1);
        }
        if (this.initiated === true) {
            console.error("Trying to re-init the transport, please avoid");
            return this;
        }
        this.initiated = true;
        var AgentNode = require('@pm2/agent-node');
        this.logger('Init new transport service');
        this.config = config;
        this.process = {
            axm_actions: [],
            axm_options: {},
            axm_monitor: {}
        };
        this.agent = new AgentNode(this.config, this.process);
        if (this.agent instanceof Error) {
            throw this.agent;
        }
        this.agent.sendLogs = config.sendLogs || false;
        this.traceAggregator = new transactionAggregator_1.TransactionAggregator();
        this.traceAggregator.init();
        this.traceAggregator.on('packet', function (packet) {
            _this.send('axm:transaction', packet);
        });
        this.agent.start();
        this.agent.transport.on('**', function (data) {
            _this.logger("Received reverse message from websocket transport");
            _this.emit('data', data);
        });
        this.logger('Agent launched');
        return this;
    };
    WebsocketTransport.prototype.setMetrics = function (metrics) {
        return this.process.axm_monitor = metrics.reduce(function (object, metric) {
            if (typeof metric.name !== 'string')
                return object;
            object[metric.name] = {
                historic: metric.historic,
                unit: metric.unit,
                type: metric.id,
                value: metric.value
            };
            return object;
        }, {});
    };
    WebsocketTransport.prototype.addAction = function (action) {
        this.logger("Add action: " + action.name + ":" + action.type);
        var serializedAction = {
            action_name: action.name,
            action_type: action.type,
            arity: action.arity,
            opts: action.opts
        };
        this.process.axm_actions.push(serializedAction);
    };
    WebsocketTransport.prototype.setOptions = function (options) {
        this.logger("Set options: [" + Object.keys(options).join(',') + "]");
        return this.process.axm_options = Object.assign(this.process.axm_options, options);
    };
    WebsocketTransport.prototype.getFormattedPayload = function (channel, payload) {
        switch (channel) {
            case 'axm:reply':
                return { data: payload };
            case 'process:exception':
                return { data: payload };
            case 'human:event': {
                var name = payload.__name;
                payload.__name = undefined;
                return { name: name, data: payload };
            }
        }
        return payload;
    };
    WebsocketTransport.prototype.send = function (channel, payload) {
        if (channel === 'axm:trace' && this.traceAggregator !== undefined) {
            return this.traceAggregator.aggregate(payload);
        }
        return this.agent.send(channel, this.getFormattedPayload(channel, payload)) ? 0 : -1;
    };
    WebsocketTransport.prototype.destroy = function () {
        this.agent.transport.disconnect();
        if (this.traceAggregator !== undefined) {
            this.traceAggregator.destroy();
        }
        this.logger('destroy');
    };
    WebsocketTransport.prototype.removeListener = function () {
        return this.agent.transport.removeListener.apply(this, arguments);
    };
    WebsocketTransport.prototype.removeAllListeners = function () {
        return this.agent.transport.removeAllListeners.apply(this, arguments);
    };
    WebsocketTransport.prototype.on = function () {
        return this.agent.transport.on.apply(this, arguments);
    };
    return WebsocketTransport;
}(eventemitter2_1.EventEmitter2));
exports.WebsocketTransport = WebsocketTransport;
