// @pm2/io/build/main/featureManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var notify_1 = require("./features/notify");
var profiling_1 = require("./features/profiling");
var eventLoopInspector_1 = require("./features/eventLoopInspector");
var events_1 = require("./features/events");
var metrics_1 = require("./features/metrics");
var Debug = require("debug");
var tracing_1 = require("./features/tracing");
function getObjectAtPath(context, path) {
    if (path.indexOf('.') === -1 && path.indexOf('[') === -1) {
        return context[path];
    }
    var crumbs = path.split(/\.|\[|\]/g);
    var i = -1;
    var len = crumbs.length;
    var result;
    while (++i < len) {
        if (i === 0)
            result = context;
        if (!crumbs[i])
            continue;
        if (result === undefined)
            break;
        result = result[crumbs[i]];
    }
    return result;
}
exports.getObjectAtPath = getObjectAtPath;
var AvailableFeature = (function () {
    function AvailableFeature() {
    }
    return AvailableFeature;
}());
var availablesFeatures = [
    {
        name: 'notify',
        optionsPath: '.',
        module: notify_1.NotifyFeature
    },
    {
        name: 'profiler',
        optionsPath: 'profiling',
        module: profiling_1.ProfilingFeature
    },
    {
        name: 'eventLoopInspector',
        optionsPath: 'actions.eventLoopDump',
        module: eventLoopInspector_1.EventLoopInspectorFeature
    },
    {
        name: 'events',
        module: events_1.EventsFeature
    },
    {
        name: 'metrics',
        optionsPath: 'metrics',
        module: metrics_1.MetricsFeature
    },
    {
        name: 'tracing',
        optionsPath: 'tracing',
        module: tracing_1.TracingFeature
    }
];
var FeatureManager = (function () {
    function FeatureManager() {
        this.logger = Debug('axm:features');
    }
    FeatureManager.prototype.init = function (options) {
        var e_1, _a;
        try {
            for (var availablesFeatures_1 = tslib_1.__values(availablesFeatures), availablesFeatures_1_1 = availablesFeatures_1.next(); !availablesFeatures_1_1.done; availablesFeatures_1_1 = availablesFeatures_1.next()) {
                var availableFeature = availablesFeatures_1_1.value;
                this.logger("Creating feature " + availableFeature.name);
                var feature = new availableFeature.module();
                var config = undefined;
                if (typeof availableFeature.optionsPath !== 'string') {
                    config = {};
                }
                else if (availableFeature.optionsPath === '.') {
                    config = options;
                }
                else {
                    config = getObjectAtPath(options, availableFeature.optionsPath);
                }
                this.logger("Init feature " + availableFeature.name);
                feature.init(config);
                availableFeature.instance = feature;
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (availablesFeatures_1_1 && !availablesFeatures_1_1.done && (_a = availablesFeatures_1.return)) _a.call(availablesFeatures_1);
            }
            finally { if (e_1) throw e_1.error; }
        }
    };
    FeatureManager.prototype.get = function (name) {
        var feature = availablesFeatures.find(function (feature) { return feature.name === name; });
        if (feature === undefined || feature.instance === undefined) {
            throw new Error("Tried to call feature " + name + " which doesn't exist or wasn't initiated");
        }
        return feature.instance;
    };
    FeatureManager.prototype.destroy = function () {
        var e_2, _a;
        try {
            for (var availablesFeatures_2 = tslib_1.__values(availablesFeatures), availablesFeatures_2_1 = availablesFeatures_2.next(); !availablesFeatures_2_1.done; availablesFeatures_2_1 = availablesFeatures_2.next()) {
                var availableFeature = availablesFeatures_2_1.value;
                if (availableFeature.instance === undefined)
                    continue;
                this.logger("Destroy feature " + availableFeature.name);
                availableFeature.instance.destroy();
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (availablesFeatures_2_1 && !availablesFeatures_2_1.done && (_a = availablesFeatures_2.return)) _a.call(availablesFeatures_2);
            }
            finally { if (e_2) throw e_2.error; }
        }
    };
    return FeatureManager;
}());
exports.FeatureManager = FeatureManager;
var FeatureConfig = (function () {
    function FeatureConfig() {
    }
    return FeatureConfig;
}());
exports.FeatureConfig = FeatureConfig;
