// @pm2/io/build/main/pmx.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var configuration_1 = require("./configuration");
var debug_1 = require("debug");
var serviceManager_1 = require("./serviceManager");
var transport_1 = require("./services/transport");
var featureManager_1 = require("./featureManager");
var actions_1 = require("./services/actions");
var metrics_1 = require("./services/metrics");
var constants_1 = require("./constants");
var runtimeStats_1 = require("./services/runtimeStats");
var entrypoint_1 = require("./features/entrypoint");
var IOConfig = (function () {
    function IOConfig() {
        this.catchExceptions = true;
        this.profiling = true;
        this.tracing = false;
        this.standalone = false;
    }
    return IOConfig;
}());
exports.IOConfig = IOConfig;
exports.defaultConfig = {
    catchExceptions: true,
    profiling: true,
    metrics: {
        v8: true,
        network: false,
        eventLoop: true,
        runtime: true,
        http: true
    },
    standalone: false,
    tracing: false
};
var PMX = (function () {
    function PMX() {
        this.featureManager = new featureManager_1.FeatureManager();
        this.transport = null;
        this.actionService = null;
        this.metricService = null;
        this.runtimeStatsService = null;
        this.logger = debug_1.default('axm:main');
        this.initialized = false;
        this.Entrypoint = entrypoint_1.Entrypoint;
    }
    PMX.prototype.init = function (config) {
        var callsite = (new Error().stack || '').split('\n')[2];
        if (callsite && callsite.length > 0) {
            this.logger("init from " + callsite);
        }
        if (this.initialized === true) {
            this.logger("Calling init but was already the case, destroying and recreating");
            this.destroy();
        }
        if (config === undefined) {
            config = exports.defaultConfig;
        }
        this.transport = transport_1.createTransport(config.standalone === true ? 'websocket' : 'ipc', config.apmOptions);
        serviceManager_1.ServiceManager.set('transport', this.transport);
        if (constants_1.canUseInspector()) {
            var Inspector = require('./services/inspector');
            var inspectorService = new Inspector();
            inspectorService.init();
            serviceManager_1.ServiceManager.set('inspector', inspectorService);
        }
        this.actionService = new actions_1.ActionService();
        this.actionService.init();
        serviceManager_1.ServiceManager.set('actions', this.actionService);
        this.metricService = new metrics_1.MetricService();
        this.metricService.init();
        serviceManager_1.ServiceManager.set('metrics', this.metricService);
        this.runtimeStatsService = new runtimeStats_1.RuntimeStatsService();
        this.runtimeStatsService.init();
        if (this.runtimeStatsService.isEnabled()) {
            serviceManager_1.ServiceManager.set('runtimeStats', this.runtimeStatsService);
        }
        this.featureManager.init(config);
        configuration_1.default.init(config);
        this.initialConfig = config;
        this.initialized = true;
        return this;
    };
    PMX.prototype.destroy = function () {
        this.logger('destroy');
        this.featureManager.destroy();
        if (this.actionService !== null) {
            this.actionService.destroy();
        }
        if (this.transport !== null) {
            this.transport.destroy();
        }
        if (this.metricService !== null) {
            this.metricService.destroy();
        }
        if (this.runtimeStatsService !== null) {
            this.runtimeStatsService.destroy();
        }
        var inspectorService = serviceManager_1.ServiceManager.get('inspector');
        if (inspectorService !== undefined) {
            inspectorService.destroy();
        }
    };
    PMX.prototype.getConfig = function () {
        return this.initialConfig;
    };
    PMX.prototype.notifyError = function (error, context) {
        var notify = this.featureManager.get('notify');
        return notify.notifyError(error, context);
    };
    PMX.prototype.metrics = function (metric) {
        var e_1, _a;
        var res = [];
        if (metric === undefined || metric === null) {
            console.error("Received empty metric to create");
            console.trace();
            return [];
        }
        var metrics = !Array.isArray(metric) ? [metric] : metric;
        try {
            for (var metrics_2 = tslib_1.__values(metrics), metrics_2_1 = metrics_2.next(); !metrics_2_1.done; metrics_2_1 = metrics_2.next()) {
                var metric_1 = metrics_2_1.value;
                if (typeof metric_1.name !== 'string') {
                    console.error("Trying to create a metrics without a name", metric_1);
                    console.trace();
                    res.push({});
                    continue;
                }
                if (metric_1.type === undefined) {
                    metric_1.type = metrics_1.MetricType.gauge;
                }
                switch (metric_1.type) {
                    case metrics_1.MetricType.counter: {
                        res.push(this.counter(metric_1));
                        continue;
                    }
                    case metrics_1.MetricType.gauge: {
                        res.push(this.gauge(metric_1));
                        continue;
                    }
                    case metrics_1.MetricType.histogram: {
                        res.push(this.histogram(metric_1));
                        continue;
                    }
                    case metrics_1.MetricType.meter: {
                        res.push(this.meter(metric_1));
                        continue;
                    }
                    case metrics_1.MetricType.metric: {
                        res.push(this.gauge(metric_1));
                        continue;
                    }
                    default: {
                        console.error("Invalid metric type " + metric_1.type + " for metric " + metric_1.name);
                        console.trace();
                        res.push({});
                        continue;
                    }
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (metrics_2_1 && !metrics_2_1.done && (_a = metrics_2.return)) _a.call(metrics_2);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return res;
    };
    PMX.prototype.histogram = function (config) {
        if (typeof config === 'string') {
            config = {
                name: config,
                measurement: metrics_1.MetricMeasurements.mean
            };
        }
        if (this.metricService === null) {
            return console.trace("Tried to register a metric without initializing @pm2/io");
        }
        return this.metricService.histogram(config);
    };
    PMX.prototype.metric = function (config) {
        if (typeof config === 'string') {
            config = {
                name: config
            };
        }
        if (this.metricService === null) {
            return console.trace("Tried to register a metric without initializing @pm2/io");
        }
        return this.metricService.metric(config);
    };
    PMX.prototype.gauge = function (config) {
        if (typeof config === 'string') {
            config = {
                name: config
            };
        }
        if (this.metricService === null) {
            return console.trace("Tried to register a metric without initializing @pm2/io");
        }
        return this.metricService.metric(config);
    };
    PMX.prototype.counter = function (config) {
        if (typeof config === 'string') {
            config = {
                name: config
            };
        }
        if (this.metricService === null) {
            return console.trace("Tried to register a metric without initializing @pm2/io");
        }
        return this.metricService.counter(config);
    };
    PMX.prototype.meter = function (config) {
        if (typeof config === 'string') {
            config = {
                name: config
            };
        }
        if (this.metricService === null) {
            return console.trace("Tried to register a metric without initializing @pm2/io");
        }
        return this.metricService.meter(config);
    };
    PMX.prototype.action = function (name, opts, fn) {
        if (typeof name === 'object') {
            var tmp = name;
            name = tmp.name;
            opts = tmp.options;
            fn = tmp.action;
        }
        if (this.actionService === null) {
            return console.trace("Tried to register a action without initializing @pm2/io");
        }
        return this.actionService.registerAction(name, opts, fn);
    };
    PMX.prototype.onExit = function (callback) {
        if (typeof callback === 'function') {
            var onExit = require('signal-exit');
            return onExit(callback);
        }
    };
    PMX.prototype.emit = function (name, data) {
        var events = this.featureManager.get('events');
        return events.emit(name, data);
    };
    PMX.prototype.initModule = function (opts, cb) {
        if (!opts)
            opts = {};
        if (opts.reference) {
            opts.name = opts.reference;
            delete opts.reference;
        }
        opts = Object.assign({
            widget: {}
        }, opts);
        opts.widget = Object.assign({
            type: 'generic',
            logo: 'https://app.keymetrics.io/img/logo/keymetrics-300.png',
            theme: ['#111111', '#1B2228', '#807C7C', '#807C7C']
        }, opts.widget);
        opts.isModule = true;
        opts = configuration_1.default.init(opts);
        return typeof cb === 'function' ? cb(null, opts) : opts;
    };
    PMX.prototype.expressErrorHandler = function () {
        var notify = this.featureManager.get('notify');
        return notify.expressErrorHandler();
    };
    PMX.prototype.koaErrorHandler = function () {
        var notify = this.featureManager.get('notify');
        return notify.koaErrorHandler();
    };
    return PMX;
}());
exports.default = PMX;
