// @pm2/io/build/main/metrics/eventLoopMetrics.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
var metrics_1 = require("../services/metrics");
var serviceManager_1 = require("../serviceManager");
var Debug = require("debug");
var histogram_1 = require("../utils/metrics/histogram");
var EventLoopMetricOption = (function () {
    function EventLoopMetricOption() {
    }
    return EventLoopMetricOption;
}());
exports.EventLoopMetricOption = EventLoopMetricOption;
var defaultOptions = {
    eventLoopActive: true,
    eventLoopDelay: true
};
var EventLoopHandlesRequestsMetric = (function () {
    function EventLoopHandlesRequestsMetric() {
        this.logger = Debug('axm:features:metrics:eventloop');
        this.delayLoopInterval = 1000;
    }
    EventLoopHandlesRequestsMetric.prototype.init = function (config) {
        var _this = this;
        if (config === false)
            return;
        if (config === undefined) {
            config = defaultOptions;
        }
        if (config === true) {
            config = defaultOptions;
        }
        this.metricService = serviceManager_1.ServiceManager.get('metrics');
        if (this.metricService === undefined)
            return this.logger('Failed to load metric service');
        this.logger('init');
        if (typeof process._getActiveRequests === 'function' && config.eventLoopActive === true) {
            var requestMetric_1 = this.metricService.metric({
                name: 'Active requests',
                id: 'internal/libuv/requests',
                historic: true
            });
            this.requestTimer = setInterval(function (_) {
                requestMetric_1.set(process._getActiveRequests().length);
            }, 1000);
            this.requestTimer.unref();
        }
        if (typeof process._getActiveHandles === 'function' && config.eventLoopActive === true) {
            var handleMetric_1 = this.metricService.metric({
                name: 'Active handles',
                id: 'internal/libuv/handles',
                historic: true
            });
            this.handleTimer = setInterval(function (_) {
                handleMetric_1.set(process._getActiveHandles().length);
            }, 1000);
            this.handleTimer.unref();
        }
        if (config.eventLoopDelay === false)
            return;
        var histogram = new histogram_1.default();
        var uvLatencyp50 = {
            name: 'Event Loop Latency',
            id: 'internal/libuv/latency/p50',
            type: metrics_1.MetricType.histogram,
            historic: true,
            implementation: histogram,
            handler: function () {
                var percentiles = this.implementation.percentiles([0.5]);
                if (percentiles[0.5] === null)
                    return null;
                return percentiles[0.5].toFixed(2);
            },
            unit: 'ms'
        };
        var uvLatencyp95 = {
            name: 'Event Loop Latency p95',
            id: 'internal/libuv/latency/p95',
            type: metrics_1.MetricType.histogram,
            historic: true,
            implementation: histogram,
            handler: function () {
                var percentiles = this.implementation.percentiles([0.95]);
                if (percentiles[0.95] === null)
                    return null;
                return percentiles[0.95].toFixed(2);
            },
            unit: 'ms'
        };
        this.metricService.registerMetric(uvLatencyp50);
        this.metricService.registerMetric(uvLatencyp95);
        this.runtimeStatsService = serviceManager_1.ServiceManager.get('runtimeStats');
        if (this.runtimeStatsService === undefined) {
            this.logger('runtimeStats module not found, fallbacking into pure js method');
            var oldTime_1 = process.hrtime();
            this.delayTimer = setInterval(function () {
                var newTime = process.hrtime();
                var delay = (newTime[0] - oldTime_1[0]) * 1e3 + (newTime[1] - oldTime_1[1]) / 1e6 - _this.delayLoopInterval;
                oldTime_1 = newTime;
                histogram.update(delay);
            }, this.delayLoopInterval);
            this.delayTimer.unref();
        }
        else {
            this.logger('using runtimeStats module as data source for event loop latency');
            this.handle = function (stats) {
                if (typeof stats !== 'object' || !Array.isArray(stats.ticks))
                    return;
                stats.ticks.forEach(function (tick) {
                    histogram.update(tick);
                });
            };
            this.runtimeStatsService.on('data', this.handle);
        }
    };
    EventLoopHandlesRequestsMetric.prototype.destroy = function () {
        if (this.requestTimer !== undefined) {
            clearInterval(this.requestTimer);
        }
        if (this.handleTimer !== undefined) {
            clearInterval(this.handleTimer);
        }
        if (this.delayTimer !== undefined) {
            clearInterval(this.delayTimer);
        }
        if (this.runtimeStatsService !== undefined) {
            this.runtimeStatsService.removeListener('data', this.handle);
        }
        this.logger('destroy');
    };
    return EventLoopHandlesRequestsMetric;
}());
exports.default = EventLoopHandlesRequestsMetric;
