// @pm2/io/build/main/metrics/httpMetrics.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
var shimmer = require("shimmer");
var debug_1 = require("debug");
var configuration_1 = require("../configuration");
var serviceManager_1 = require("../serviceManager");
var histogram_1 = require("../utils/metrics/histogram");
var Module = require("module");
var metrics_1 = require("../services/metrics");
var HttpMetricsConfig = (function () {
    function HttpMetricsConfig() {
    }
    return HttpMetricsConfig;
}());
exports.HttpMetricsConfig = HttpMetricsConfig;
var HttpMetrics = (function () {
    function HttpMetrics() {
        this.defaultConf = {
            http: true
        };
        this.metrics = new Map();
        this.logger = debug_1.default('axm:features:metrics:http');
        this.modules = {};
    }
    HttpMetrics.prototype.init = function (config) {
        if (config === undefined)
            return;
        if (config === false)
            return;
        if (typeof config !== 'object') {
            config = this.defaultConf;
        }
        this.logger('init');
        configuration_1.default.configureModule({
            latency: true
        });
        this.metricService = serviceManager_1.ServiceManager.get('metrics');
        if (this.metricService === undefined)
            return this.logger("Failed to load metric service");
        this.logger('hooking to require');
        this.hookRequire();
    };
    HttpMetrics.prototype.registerHttpMetric = function () {
        if (this.metricService === undefined)
            return this.logger("Failed to load metric service");
        var histogram = new histogram_1.default();
        var p50 = {
            name: "HTTP Mean Latency",
            id: 'internal/http/builtin/latency/p50',
            type: metrics_1.MetricType.histogram,
            historic: true,
            implementation: histogram,
            unit: 'ms',
            handler: function () {
                var percentiles = histogram.percentiles([0.5]);
                return percentiles[0.5];
            }
        };
        var p95 = {
            name: "HTTP P95 Latency",
            id: 'internal/http/builtin/latency/p95',
            type: metrics_1.MetricType.histogram,
            historic: true,
            implementation: histogram,
            handler: function () {
                var percentiles = histogram.percentiles([0.95]);
                return percentiles[0.95];
            },
            unit: 'ms'
        };
        var meter = {
            name: 'HTTP',
            historic: true,
            id: 'internal/http/builtin/reqs',
            unit: 'req/min'
        };
        this.metricService.registerMetric(p50);
        this.metricService.registerMetric(p95);
        this.metrics.set('http.latency', histogram);
        this.metrics.set('http.meter', this.metricService.meter(meter));
    };
    HttpMetrics.prototype.registerHttpsMetric = function () {
        if (this.metricService === undefined)
            return this.logger("Failed to load metric service");
        var histogram = new histogram_1.default();
        var p50 = {
            name: "HTTPS Mean Latency",
            id: 'internal/https/builtin/latency/p50',
            type: metrics_1.MetricType.histogram,
            historic: true,
            implementation: histogram,
            unit: 'ms',
            handler: function () {
                var percentiles = histogram.percentiles([0.5]);
                return percentiles[0.5];
            }
        };
        var p95 = {
            name: "HTTPS P95 Latency",
            id: 'internal/https/builtin/latency/p95',
            type: metrics_1.MetricType.histogram,
            historic: true,
            implementation: histogram,
            handler: function () {
                var percentiles = histogram.percentiles([0.95]);
                return percentiles[0.95];
            },
            unit: 'ms'
        };
        var meter = {
            name: 'HTTPS',
            historic: true,
            id: 'internal/https/builtin/reqs',
            unit: 'req/min'
        };
        this.metricService.registerMetric(p50);
        this.metricService.registerMetric(p95);
        this.metrics.set('https.latency', histogram);
        this.metrics.set('https.meter', this.metricService.meter(meter));
    };
    HttpMetrics.prototype.destroy = function () {
        if (this.modules.http !== undefined) {
            this.logger('unwraping http module');
            shimmer.unwrap(this.modules.http, 'emit');
        }
        if (this.modules.https !== undefined) {
            this.logger('unwraping https module');
            shimmer.unwrap(this.modules.https, 'emit');
        }
        if (Module['_load'] && Module['_load'].__io_apm === true) {
            shimmer.unwrap(Module, '_load');
            Module['_load'].__io_apm = undefined;
        }
        this.logger('destroy');
    };
    HttpMetrics.prototype.hookHttp = function (nodule, name) {
        if (nodule.Server === undefined || nodule.Server.prototype === undefined)
            return;
        if (this.modules[name] !== undefined)
            return this.logger("Module " + name + " already hooked");
        this.logger("Hooking to " + name + " module");
        this.modules[name] = nodule.Server.prototype;
        var self = this;
        shimmer.wrap(nodule.Server.prototype, 'emit', function (original) {
            return function (event, req, res) {
                if (event !== 'request')
                    return original.apply(this, arguments);
                var meter = self.metrics.get(name + ".meter");
                if (meter !== undefined) {
                    meter.mark();
                }
                var latency = self.metrics.get(name + ".latency");
                if (latency === undefined)
                    return original.apply(this, arguments);
                if (res === undefined || res === null)
                    return original.apply(this, arguments);
                var startTime = Date.now();
                res.once('finish', function (_) {
                    latency.update(Date.now() - startTime);
                });
                return original.apply(this, arguments);
            };
        });
    };
    HttpMetrics.prototype.hookRequire = function () {
        var self = this;
        var isAlreadyWrapped = typeof Module['_load'] === 'function' && Module['_load'].__io_apm === true;
        if (isAlreadyWrapped)
            return this.logger("Module already wrapped, aborting");
        shimmer.wrap(Module, '_load', function (original) {
            return function (file) {
                var returned = original.apply(original, arguments);
                if (file === 'http' && self.modules['http'] === undefined) {
                    self.registerHttpMetric();
                    self.hookHttp(returned, file);
                }
                else if (file === 'https' && self.modules['https'] === undefined) {
                    self.registerHttpsMetric();
                    self.hookHttp(returned, file);
                }
                return returned;
            };
        });
        Module['_load'].__io_apm = true;
    };
    return HttpMetrics;
}());
exports.default = HttpMetrics;
