// @pm2/io/build/main/features/entrypoint.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var IO_KEY = Symbol.for('@pm2/io');
var Entrypoint = (function () {
    function Entrypoint() {
        var _this = this;
        try {
            this.io = global[IO_KEY].init(this.conf());
            this.onStart(function (err) {
                if (err) {
                    console.error(err);
                    process.exit(1);
                }
                _this.sensors();
                _this.events();
                _this.actuators();
                _this.io.onExit(function (code, signal) {
                    _this.onStop(err, function () {
                        _this.io.destroy();
                    }, code, signal);
                });
                if (process && process.send)
                    process.send('ready');
            });
        }
        catch (e) {
            if (this.io) {
                this.io.destroy();
            }
            throw (e);
        }
    }
    Entrypoint.prototype.events = function () {
        return;
    };
    Entrypoint.prototype.sensors = function () {
        return;
    };
    Entrypoint.prototype.actuators = function () {
        return;
    };
    Entrypoint.prototype.onStart = function (cb) {
        throw new Error('Entrypoint onStart() not specified');
    };
    Entrypoint.prototype.onStop = function (err, cb, code, signal) {
        return cb();
    };
    Entrypoint.prototype.conf = function () {
        return undefined;
    };
    return Entrypoint;
}());
exports.Entrypoint = Entrypoint;
