// @pm2/io/build/main/features/eventLoopInspector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var serviceManager_1 = require("../serviceManager");
var EventLoopInspector = require("event-loop-inspector");
var Debug = require("debug");
var EventLoopInspectorConfig = (function () {
    function EventLoopInspectorConfig() {
    }
    return EventLoopInspectorConfig;
}());
var EventLoopInspectorFeature = (function () {
    function EventLoopInspectorFeature() {
        this.logger = Debug('axm:features:actions:eventloop');
    }
    EventLoopInspectorFeature.prototype.init = function (config) {
        var _this = this;
        if (config === false)
            return;
        if (config === undefined)
            return;
        if (config === true) {
            config = { enabled: true };
        }
        if (config.enabled === false)
            return;
        this.actionService = serviceManager_1.ServiceManager.get('actions');
        if (this.actionService === undefined) {
            return this.logger('cannot expose actions as action service isnt available');
        }
        this.eventLoopInspector = EventLoopInspector(false);
        this.actionService.registerAction('km:event-loop-dump', function (cb) {
            return cb({
                success: true,
                dump: _this.eventLoopInspector.dump()
            });
        });
    };
    EventLoopInspectorFeature.prototype.destroy = function () {
        this.logger('destroy');
    };
    return EventLoopInspectorFeature;
}());
exports.EventLoopInspectorFeature = EventLoopInspectorFeature;
