// @pm2/io/build/main/features/events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var serviceManager_1 = require("../serviceManager");
var Debug = require("debug");
var EventsFeature = (function () {
    function EventsFeature() {
        this.logger = Debug('axm:features:events');
    }
    EventsFeature.prototype.init = function () {
        this.transport = serviceManager_1.ServiceManager.get('transport');
        this.logger('init');
    };
    EventsFeature.prototype.emit = function (name, data) {
        if (typeof name !== 'string') {
            console.error('event name must be a string');
            return console.trace();
        }
        if (typeof data !== 'object') {
            console.error('event data must be an object');
            return console.trace();
        }
        if (data instanceof Array) {
            console.error("event data cannot be an array");
            return console.trace();
        }
        var inflightObj = {};
        try {
            inflightObj = JSON.parse(JSON.stringify(data));
        }
        catch (err) {
            return console.log('Failed to serialize the event data', err.message);
        }
        inflightObj.__name = name;
        if (this.transport === undefined) {
            return this.logger('Failed to send event as transporter isnt available');
        }
        this.transport.send('human:event', inflightObj);
    };
    EventsFeature.prototype.destroy = function () {
        this.logger('destroy');
    };
    return EventsFeature;
}());
exports.EventsFeature = EventsFeature;
