// @pm2/io/build/main/features/profiling.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var addonProfiler_1 = require("../profilers/addonProfiler");
var inspectorProfiler_1 = require("../profilers/inspectorProfiler");
var constants_1 = require("../constants");
var Debug = require("debug");
var ProfilingConfig = (function () {
    function ProfilingConfig() {
    }
    return ProfilingConfig;
}());
exports.ProfilingConfig = ProfilingConfig;
var defaultProfilingConfig = {
    cpuJS: true,
    heapSnapshot: true,
    heapSampling: true,
    implementation: 'both'
};
var disabledProfilingConfig = {
    cpuJS: false,
    heapSnapshot: false,
    heapSampling: false,
    implementation: 'none'
};
var ProfilingFeature = (function () {
    function ProfilingFeature() {
        this.logger = Debug('axm:features:profiling');
    }
    ProfilingFeature.prototype.init = function (config) {
        if (config === true) {
            config = defaultProfilingConfig;
        }
        else if (config === false) {
            config = disabledProfilingConfig;
        }
        else if (config === undefined) {
            config = defaultProfilingConfig;
        }
        if (process.env.PM2_PROFILING_FORCE_FALLBACK === 'true') {
            config.implementation = 'addon';
        }
        if (config.implementation === undefined || config.implementation === 'both') {
            config.implementation = constants_1.canUseInspector() === true ? 'inspector' : 'addon';
        }
        switch (config.implementation) {
            case 'inspector': {
                this.logger('using inspector implementation');
                this.profiler = new inspectorProfiler_1.default();
                break;
            }
            case 'addon': {
                this.logger('using addon implementation');
                this.profiler = new addonProfiler_1.default();
                break;
            }
            default: {
                return this.logger("Invalid profiler implementation choosen: " + config.implementation);
            }
        }
        this.logger('init');
        this.profiler.init();
    };
    ProfilingFeature.prototype.destroy = function () {
        this.logger('destroy');
        if (this.profiler === undefined)
            return;
        this.profiler.destroy();
    };
    return ProfilingFeature;
}());
exports.ProfilingFeature = ProfilingFeature;
