// @pm2/io/build/main/features/tracing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var serviceManager_1 = require("../serviceManager");
var Debug = require("debug");
var configuration_1 = require("../configuration");
var PluginConfig = (function () {
    function PluginConfig() {
    }
    return PluginConfig;
}());
exports.PluginConfig = PluginConfig;
var TracerIgnoreFilter = (function () {
    function TracerIgnoreFilter() {
    }
    return TracerIgnoreFilter;
}());
exports.TracerIgnoreFilter = TracerIgnoreFilter;
var TracingConfig = (function () {
    function TracingConfig() {
    }
    return TracingConfig;
}());
exports.TracingConfig = TracingConfig;
var enabledConfig = {
    enabled: true,
    ignoreFilter: {
        url: [],
        method: ['OPTIONS']
    },
    logLevel: 1
};
var TracingFeature = (function () {
    function TracingFeature() {
        this.logger = Debug('axm:features:tracing');
    }
    TracingFeature.prototype.init = function (config) {
        var _this = this;
        if (config === undefined)
            return;
        if (config === false)
            return;
        if (config === true) {
            config = enabledConfig;
        }
        if (config.enabled === false)
            return;
        this.logger('init');
        this.transport = serviceManager_1.ServiceManager.get('transport');
        if (this.transport === undefined) {
            return this.logger("Failed to load transporter service");
        }
        var tracerModule = require('@pm2/legacy-tracing');
        if (tracerModule.get().isActive()) {
            return this.logger("Tracing already enalbed");
        }
        this.tracer = tracerModule.start(config);
        this.logger("Tracing start and correctly enabled");
        configuration_1.default.configureModule({
            tracing_enabled: true
        });
        this.traceHandler = function (data) {
            if (_this.transport === undefined)
                return;
            _this.logger('received trace data');
            _this.transport.send('axm:trace', data);
        };
        this.tracer.getBus().on('transaction', this.traceHandler);
    };
    TracingFeature.prototype.destroy = function () {
        if (this.tracer === undefined)
            return;
        this.tracer.getBus().removeAllListeners();
        if (typeof this.tracer.disable_ === 'function') {
            this.tracer.disable_();
            this.logger('stopped tracing agent');
        }
        this.logger('destroy');
    };
    return TracingFeature;
}());
exports.TracingFeature = TracingFeature;
