// @pm2/io/build/main/profilers/addonProfiler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var module_1 = require("../utils/module");
var configuration_1 = require("../configuration");
var serviceManager_1 = require("../serviceManager");
var miscellaneous_1 = require("../utils/miscellaneous");
var Debug = require("debug");
var CurrentProfile = (function () {
    function CurrentProfile() {
    }
    return CurrentProfile;
}());
var AddonProfiler = (function () {
    function AddonProfiler() {
        this.profiler = null;
        this.modules = ['v8-profiler-node8', 'v8-profiler'];
        this.currentProfile = null;
        this.logger = Debug('axm:features:profiling:addon');
    }
    AddonProfiler.prototype.init = function () {
        var e_1, _a;
        try {
            for (var _b = tslib_1.__values(this.modules), _c = _b.next(); !_c.done; _c = _b.next()) {
                var moduleName = _c.value;
                var path = module_1.default.detectModule(moduleName);
                if (path === null)
                    continue;
                var profiler = module_1.default.loadModule(moduleName);
                if (profiler instanceof Error)
                    continue;
                this.profiler = profiler;
                break;
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        if (this.profiler === null) {
            configuration_1.default.configureModule({
                heapdump: false,
                'feature.profiler.heap_snapshot': false,
                'feature.profiler.heap_sampling': false,
                'feature.profiler.cpu_js': false
            });
            return this.logger("Failed to require the profiler via addon, disabling profiling ...");
        }
        this.logger('init');
        this.actionService = serviceManager_1.ServiceManager.get('actions');
        if (this.actionService === undefined) {
            return this.logger("Fail to get action service");
        }
        this.transport = serviceManager_1.ServiceManager.get('transport');
        if (this.transport === undefined) {
            return this.logger("Fail to get transport service");
        }
        configuration_1.default.configureModule({
            heapdump: true,
            'feature.profiler.heapsnapshot': true,
            'feature.profiler.heapsampling': false,
            'feature.profiler.cpu_js': true
        });
        this.register();
    };
    AddonProfiler.prototype.register = function () {
        if (this.actionService === undefined) {
            return this.logger("Fail to get action service");
        }
        this.logger('register');
        this.actionService.registerAction('km:heapdump', this.onHeapdump.bind(this));
        this.actionService.registerAction('km:cpu:profiling:start', this.onCPUProfileStart.bind(this));
        this.actionService.registerAction('km:cpu:profiling:stop', this.onCPUProfileStop.bind(this));
    };
    AddonProfiler.prototype.destroy = function () {
        this.logger('Addon Profiler destroyed !');
        if (this.profiler === null)
            return;
        this.profiler.deleteAllProfiles();
    };
    AddonProfiler.prototype.onCPUProfileStart = function (opts, cb) {
        var _this = this;
        if (typeof cb !== 'function') {
            cb = opts;
            opts = {};
        }
        if (typeof opts !== 'object' || opts === null) {
            opts = {};
        }
        if (this.currentProfile !== null) {
            return cb({
                err: 'A profiling is already running',
                success: false
            });
        }
        this.currentProfile = new CurrentProfile();
        this.currentProfile.uuid = miscellaneous_1.default.generateUUID();
        this.currentProfile.startTime = Date.now();
        this.currentProfile.initiated = typeof opts.initiated === 'string'
            ? opts.initiated : 'manual';
        cb({ success: true, uuid: this.currentProfile.uuid });
        this.profiler.startProfiling();
        if (isNaN(parseInt(opts.timeout, 10)))
            return;
        var duration = parseInt(opts.timeout, 10);
        setTimeout(function (_) {
            _this.onCPUProfileStop(function (_) {
                return;
            });
        }, duration);
    };
    AddonProfiler.prototype.onCPUProfileStop = function (cb) {
        if (this.currentProfile === null) {
            return cb({
                err: 'No profiling are already running',
                success: false
            });
        }
        if (this.transport === undefined) {
            return cb({
                err: 'No profiling are already running',
                success: false
            });
        }
        var profile = this.profiler.stopProfiling();
        var data = JSON.stringify(profile);
        cb({ success: true, uuid: this.currentProfile.uuid });
        this.transport.send('profilings', {
            uuid: this.currentProfile.uuid,
            duration: Date.now() - this.currentProfile.startTime,
            at: this.currentProfile.startTime,
            data: data,
            dump_file_size: data.length,
            success: true,
            initiated: this.currentProfile.initiated,
            type: 'cpuprofile',
            cpuprofile: true
        });
        this.currentProfile = null;
    };
    AddonProfiler.prototype.onHeapdump = function (opts, cb) {
        var _this = this;
        if (typeof cb !== 'function') {
            cb = opts;
            opts = {};
        }
        if (typeof opts !== 'object' || opts === null) {
            opts = {};
        }
        cb({ success: true });
        setTimeout(function () {
            var startTime = Date.now();
            _this.takeSnapshot()
                .then(function (data) {
                return _this.transport.send('profilings', {
                    data: data,
                    at: startTime,
                    initiated: typeof opts.initiated === 'string' ? opts.initiated : 'manual',
                    duration: Date.now() - startTime,
                    type: 'heapdump'
                });
            }).catch(function (err) {
                return cb({
                    success: err.message,
                    err: err
                });
            });
        }, 200);
    };
    AddonProfiler.prototype.takeSnapshot = function () {
        var _this = this;
        return new Promise(function (resolve, reject) {
            var snapshot = _this.profiler.takeSnapshot();
            snapshot.export(function (err, data) {
                if (err) {
                    reject(err);
                }
                else {
                    resolve(data);
                }
                snapshot.delete();
            });
        });
    };
    return AddonProfiler;
}());
exports.default = AddonProfiler;
