// @pm2/io/build/main/profilers/inspectorProfiler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var configuration_1 = require("../configuration");
var serviceManager_1 = require("../serviceManager");
var miscellaneous_1 = require("../utils/miscellaneous");
var Debug = require("debug");
var semver = require("semver");
var CurrentProfile = (function () {
    function CurrentProfile() {
    }
    return CurrentProfile;
}());
var InspectorProfiler = (function () {
    function InspectorProfiler() {
        this.profiler = undefined;
        this.currentProfile = null;
        this.logger = Debug('axm:features:profiling:inspector');
        this.isNode11 = semver.satisfies(semver.clean(process.version), '>11.x');
    }
    InspectorProfiler.prototype.init = function () {
        this.profiler = serviceManager_1.ServiceManager.get('inspector');
        if (this.profiler === undefined) {
            configuration_1.default.configureModule({
                heapdump: false,
                'feature.profiler.heap_snapshot': false,
                'feature.profiler.heap_sampling': false,
                'feature.profiler.cpu_js': false
            });
            return console.error("Failed to require the profiler via inspector, disabling profiling ...");
        }
        this.profiler.getSession().post('Profiler.enable');
        this.profiler.getSession().post('HeapProfiler.enable');
        this.logger('init');
        this.actionService = serviceManager_1.ServiceManager.get('actions');
        if (this.actionService === undefined) {
            return this.logger("Fail to get action service");
        }
        this.transport = serviceManager_1.ServiceManager.get('transport');
        if (this.transport === undefined) {
            return this.logger("Fail to get transport service");
        }
        configuration_1.default.configureModule({
            heapdump: true,
            'feature.profiler.heapsnapshot': !this.isNode11,
            'feature.profiler.heapsampling': true,
            'feature.profiler.cpu_js': true
        });
        this.register();
    };
    InspectorProfiler.prototype.register = function () {
        if (this.actionService === undefined) {
            return this.logger("Fail to get action service");
        }
        this.logger('register');
        this.actionService.registerAction('km:heapdump', this.onHeapdump.bind(this));
        this.actionService.registerAction('km:cpu:profiling:start', this.onCPUProfileStart.bind(this));
        this.actionService.registerAction('km:cpu:profiling:stop', this.onCPUProfileStop.bind(this));
        this.actionService.registerAction('km:heap:sampling:start', this.onHeapProfileStart.bind(this));
        this.actionService.registerAction('km:heap:sampling:stop', this.onHeapProfileStop.bind(this));
    };
    InspectorProfiler.prototype.destroy = function () {
        this.logger('Inspector Profiler destroyed !');
        if (this.profiler === undefined)
            return;
        this.profiler.getSession().post('Profiler.disable');
        this.profiler.getSession().post('HeapProfiler.disable');
    };
    InspectorProfiler.prototype.onHeapProfileStart = function (opts, cb) {
        var _this = this;
        if (typeof cb !== 'function') {
            cb = opts;
            opts = {};
        }
        if (typeof opts !== 'object' || opts === null) {
            opts = {};
        }
        if (this.profiler === undefined) {
            return cb({
                err: 'Profiler not available',
                success: false
            });
        }
        if (this.currentProfile !== null) {
            return cb({
                err: 'A profiling is already running',
                success: false
            });
        }
        this.currentProfile = new CurrentProfile();
        this.currentProfile.uuid = miscellaneous_1.default.generateUUID();
        this.currentProfile.startTime = Date.now();
        this.currentProfile.initiated = typeof opts.initiated === 'string'
            ? opts.initiated : 'manual';
        cb({ success: true, uuid: this.currentProfile.uuid });
        var defaultSamplingInterval = 16384;
        this.profiler.getSession().post('HeapProfiler.startSampling', {
            samplingInterval: typeof opts.samplingInterval === 'number'
                ? opts.samplingInterval : defaultSamplingInterval
        });
        if (isNaN(parseInt(opts.timeout, 10)))
            return;
        var duration = parseInt(opts.timeout, 10);
        setTimeout(function (_) {
            _this.onHeapProfileStop(function (_) {
                return;
            });
        }, duration);
    };
    InspectorProfiler.prototype.onHeapProfileStop = function (cb) {
        var _this = this;
        if (this.currentProfile === null) {
            return cb({
                err: 'No profiling are already running',
                success: false
            });
        }
        if (this.profiler === undefined) {
            return cb({
                err: 'Profiler not available',
                success: false
            });
        }
        cb({ success: true, uuid: this.currentProfile.uuid });
        this.profiler.getSession().post('HeapProfiler.stopSampling', function (_, _a) {
            var profile = _a.profile;
            if (_this.currentProfile === null)
                return;
            if (_this.transport === undefined)
                return;
            var data = JSON.stringify(profile);
            _this.transport.send('profilings', {
                uuid: _this.currentProfile.uuid,
                duration: Date.now() - _this.currentProfile.startTime,
                at: _this.currentProfile.startTime,
                data: data,
                success: true,
                initiated: _this.currentProfile.initiated,
                type: 'heapprofile',
                heapprofile: true
            });
            _this.currentProfile = null;
        });
    };
    InspectorProfiler.prototype.onCPUProfileStart = function (opts, cb) {
        var _this = this;
        if (typeof cb !== 'function') {
            cb = opts;
            opts = {};
        }
        if (typeof opts !== 'object' || opts === null) {
            opts = {};
        }
        if (this.profiler === undefined) {
            return cb({
                err: 'Profiler not available',
                success: false
            });
        }
        if (this.currentProfile !== null) {
            return cb({
                err: 'A profiling is already running',
                success: false
            });
        }
        this.currentProfile = new CurrentProfile();
        this.currentProfile.uuid = miscellaneous_1.default.generateUUID();
        this.currentProfile.startTime = Date.now();
        this.currentProfile.initiated = typeof opts.initiated === 'string'
            ? opts.initiated : 'manual';
        cb({ success: true, uuid: this.currentProfile.uuid });
        if (process.hasOwnProperty('_startProfilerIdleNotifier') === true) {
            process._startProfilerIdleNotifier();
        }
        this.profiler.getSession().post('Profiler.start');
        if (isNaN(parseInt(opts.timeout, 10)))
            return;
        var duration = parseInt(opts.timeout, 10);
        setTimeout(function (_) {
            _this.onCPUProfileStop(function (_) {
                return;
            });
        }, duration);
    };
    InspectorProfiler.prototype.onCPUProfileStop = function (cb) {
        var _this = this;
        if (this.currentProfile === null) {
            return cb({
                err: 'No profiling are already running',
                success: false
            });
        }
        if (this.profiler === undefined) {
            return cb({
                err: 'Profiler not available',
                success: false
            });
        }
        cb({ success: true, uuid: this.currentProfile.uuid });
        if (process.hasOwnProperty('_stopProfilerIdleNotifier') === true) {
            process._stopProfilerIdleNotifier();
        }
        this.profiler.getSession().post('Profiler.stop', function (_, res) {
            if (_this.currentProfile === null)
                return;
            if (_this.transport === undefined)
                return;
            var profile = res.profile;
            var data = JSON.stringify(profile);
            _this.transport.send('profilings', {
                uuid: _this.currentProfile.uuid,
                duration: Date.now() - _this.currentProfile.startTime,
                at: _this.currentProfile.startTime,
                data: data,
                success: true,
                initiated: _this.currentProfile.initiated,
                type: 'cpuprofile',
                cpuprofile: true
            });
            _this.currentProfile = null;
        });
    };
    InspectorProfiler.prototype.onHeapdump = function (opts, cb) {
        var _this = this;
        if (typeof cb !== 'function') {
            cb = opts;
            opts = {};
        }
        if (typeof opts !== 'object' || opts === null) {
            opts = {};
        }
        if (this.profiler === undefined) {
            return cb({
                err: 'Profiler not available',
                success: false
            });
        }
        cb({ success: true });
        setTimeout(function () {
            var startTime = Date.now();
            _this.takeSnapshot()
                .then(function (data) {
                return _this.transport.send('profilings', {
                    data: data,
                    at: startTime,
                    initiated: typeof opts.initiated === 'string' ? opts.initiated : 'manual',
                    duration: Date.now() - startTime,
                    type: 'heapdump'
                });
            }).catch(function (err) {
                return cb({
                    success: err.message,
                    err: err
                });
            });
        }, 200);
    };
    InspectorProfiler.prototype.takeSnapshot = function () {
        var _this = this;
        return new Promise(function (resolve, reject) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
            var chunks, chunkHandler;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (this.profiler === undefined)
                            return [2, reject(new Error("Profiler not available"))];
                        chunks = [];
                        chunkHandler = function (raw) {
                            var data = raw.params;
                            chunks.push(data.chunk);
                        };
                        this.profiler.getSession().on('HeapProfiler.addHeapSnapshotChunk', chunkHandler);
                        return [4, this.profiler.getSession().post('HeapProfiler.takeHeapSnapshot', {
                                reportProgress: false
                            })];
                    case 1:
                        _a.sent();
                        this.profiler.getSession().removeListener('HeapProfiler.addHeapSnapshotChunk', chunkHandler);
                        return [2, resolve(chunks.join(''))];
                }
            });
        }); });
    };
    return InspectorProfiler;
}());
exports.default = InspectorProfiler;
