// @pm2/io/build/main/services/actions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var serviceManager_1 = require("../serviceManager");
var Debug = require("debug");
var Action = (function () {
    function Action() {
    }
    return Action;
}());
exports.Action = Action;
var ActionService = (function () {
    function ActionService() {
        this.timer = undefined;
        this.transport = undefined;
        this.actions = new Map();
        this.logger = Debug('axm:services:actions');
    }
    ActionService.prototype.listener = function (data) {
        var _this = this;
        this.logger("Received new message from reverse");
        if (!data)
            return false;
        var actionName = data.msg ? data.msg : data.action_name ? data.action_name : data;
        var action = this.actions.get(actionName);
        if (typeof action !== 'object') {
            return this.logger("Received action " + actionName + " but failed to find the implementation");
        }
        if (!action.isScoped) {
            this.logger("Succesfully called custom action " + action.name + " with arity " + action.handler.length);
            if (action.handler.length === 2) {
                var params = {};
                if (typeof data === 'object') {
                    params = data.opts;
                }
                return action.handler(params, action.callback);
            }
            return action.handler(action.callback);
        }
        if (data.uuid === undefined) {
            return this.logger("Received scoped action " + action.name + " but without uuid");
        }
        var stream = {
            send: function (dt) {
                _this.transport.send('axm:scoped_action:stream', {
                    data: dt,
                    uuid: data.uuid,
                    action_name: actionName
                });
            },
            error: function (dt) {
                _this.transport.send('axm:scoped_action:error', {
                    data: dt,
                    uuid: data.uuid,
                    action_name: actionName
                });
            },
            end: function (dt) {
                _this.transport.send('axm:scoped_action:end', {
                    data: dt,
                    uuid: data.uuid,
                    action_name: actionName
                });
            }
        };
        this.logger("Succesfully called scoped action " + action.name);
        return action.handler(data.opts || {}, stream);
    };
    ActionService.prototype.init = function () {
        this.transport = serviceManager_1.ServiceManager.get('transport');
        if (this.transport === undefined) {
            return this.logger("Failed to load transport service");
        }
        this.actions.clear();
        this.transport.on('data', this.listener.bind(this));
    };
    ActionService.prototype.destroy = function () {
        if (this.timer !== undefined) {
            clearInterval(this.timer);
        }
        if (this.transport !== undefined) {
            this.transport.removeListener('data', this.listener.bind(this));
        }
    };
    ActionService.prototype.registerAction = function (actionName, opts, handler) {
        var _this = this;
        if (typeof opts === 'function') {
            handler = opts;
            opts = undefined;
        }
        if (typeof actionName !== 'string') {
            console.error("You must define an name when registering an action");
            return;
        }
        if (typeof handler !== 'function') {
            console.error("You must define an callback when registering an action");
            return;
        }
        if (this.transport === undefined) {
            return this.logger("Failed to load transport service");
        }
        var type = 'custom';
        if (actionName.indexOf('km:') === 0 || actionName.indexOf('internal:') === 0) {
            type = 'internal';
        }
        var reply = function (data) {
            _this.transport.send('axm:reply', {
                at: new Date().getTime(),
                action_name: actionName,
                return: data
            });
        };
        var action = {
            name: actionName,
            callback: reply,
            handler: handler,
            type: type,
            isScoped: false,
            arity: handler.length,
            opts: opts
        };
        this.logger("Succesfully registered custom action " + action.name);
        this.actions.set(actionName, action);
        this.transport.addAction(action);
    };
    ActionService.prototype.scopedAction = function (actionName, handler) {
        if (typeof actionName !== 'string') {
            console.error("You must define an name when registering an action");
            return -1;
        }
        if (typeof handler !== 'function') {
            console.error("You must define an callback when registering an action");
            return -1;
        }
        if (this.transport === undefined) {
            return this.logger("Failed to load transport service");
        }
        var action = {
            name: actionName,
            handler: handler,
            type: 'scoped',
            isScoped: true,
            arity: handler.length,
            opts: null
        };
        this.logger("Succesfully registered scoped action " + action.name);
        this.actions.set(actionName, action);
        this.transport.addAction(action);
    };
    return ActionService;
}());
exports.ActionService = ActionService;
