// @pm2/io/build/main/services/inspector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var inspector = require("inspector");
var debug_1 = require("debug");
var InspectorService = (function () {
    function InspectorService() {
        this.session = null;
        this.logger = debug_1.default('axm:services:inspector');
    }
    InspectorService.prototype.init = function () {
        this.logger("Creating new inspector session");
        this.session = new inspector.Session();
        this.session.connect();
        this.logger('Connected to inspector');
        this.session.post('Profiler.enable');
        this.session.post('HeapProfiler.enable');
        return this.session;
    };
    InspectorService.prototype.getSession = function () {
        if (this.session === null) {
            this.session = this.init();
            return this.session;
        }
        else {
            return this.session;
        }
    };
    InspectorService.prototype.destroy = function () {
        if (this.session !== null) {
            this.session.post('Profiler.disable');
            this.session.post('HeapProfiler.disable');
            this.session.disconnect();
            this.session = null;
        }
        else {
            this.logger('No open session');
        }
    };
    return InspectorService;
}());
exports.InspectorService = InspectorService;
module.exports = InspectorService;
