// @pm2/io/build/main/services/metrics.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var meter_1 = require("../utils/metrics/meter");
var counter_1 = require("../utils/metrics/counter");
var histogram_1 = require("../utils/metrics/histogram");
var serviceManager_1 = require("../serviceManager");
var constants_1 = require("../constants");
var Debug = require("debug");
var gauge_1 = require("../utils/metrics/gauge");
var MetricType;
(function (MetricType) {
    MetricType["meter"] = "meter";
    MetricType["histogram"] = "histogram";
    MetricType["counter"] = "counter";
    MetricType["gauge"] = "gauge";
    MetricType["metric"] = "metric";
})(MetricType = exports.MetricType || (exports.MetricType = {}));
var MetricMeasurements;
(function (MetricMeasurements) {
    MetricMeasurements["min"] = "min";
    MetricMeasurements["max"] = "max";
    MetricMeasurements["sum"] = "sum";
    MetricMeasurements["count"] = "count";
    MetricMeasurements["variance"] = "variance";
    MetricMeasurements["mean"] = "mean";
    MetricMeasurements["stddev"] = "stddev";
    MetricMeasurements["median"] = "median";
    MetricMeasurements["p75"] = "p75";
    MetricMeasurements["p95"] = "p95";
    MetricMeasurements["p99"] = "p99";
    MetricMeasurements["p999"] = "p999";
})(MetricMeasurements = exports.MetricMeasurements || (exports.MetricMeasurements = {}));
var Metric = (function () {
    function Metric() {
    }
    return Metric;
}());
exports.Metric = Metric;
var MetricBulk = (function (_super) {
    tslib_1.__extends(MetricBulk, _super);
    function MetricBulk() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return MetricBulk;
}(Metric));
exports.MetricBulk = MetricBulk;
var HistogramOptions = (function (_super) {
    tslib_1.__extends(HistogramOptions, _super);
    function HistogramOptions() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return HistogramOptions;
}(Metric));
exports.HistogramOptions = HistogramOptions;
var MetricService = (function () {
    function MetricService() {
        this.metrics = new Map();
        this.timer = null;
        this.transport = null;
        this.logger = Debug('axm:services:metrics');
    }
    MetricService.prototype.init = function () {
        var _this = this;
        this.transport = serviceManager_1.ServiceManager.get('transport');
        if (this.transport === null)
            return this.logger('Failed to init metrics service cause no transporter');
        this.logger('init');
        this.timer = setInterval(function () {
            var e_1, _a;
            if (_this.transport === null)
                return _this.logger('Abort metrics update since transport is not available');
            _this.logger('refreshing metrics value');
            try {
                for (var _b = tslib_1.__values(_this.metrics.values()), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var metric = _c.value;
                    metric.value = metric.handler();
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
            _this.logger('sending update metrics value to transporter');
            var metricsToSend = Array.from(_this.metrics.values())
                .filter(function (metric) {
                if (metric === null || metric === undefined)
                    return false;
                if (metric.value === undefined || metric.value === null)
                    return false;
                var isNumber = typeof metric.value === 'number';
                var isString = typeof metric.value === 'string';
                var isValidNumber = !isNaN(metric.value);
                return isString || (isNumber && isValidNumber);
            });
            _this.transport.setMetrics(metricsToSend);
        }, constants_1.default.METRIC_INTERVAL);
        this.timer.unref();
    };
    MetricService.prototype.registerMetric = function (metric) {
        if (typeof metric.name !== 'string') {
            console.error("Invalid metric name declared: " + metric.name);
            return console.trace();
        }
        else if (typeof metric.type !== 'string') {
            console.error("Invalid metric type declared: " + metric.type);
            return console.trace();
        }
        else if (typeof metric.handler !== 'function') {
            console.error("Invalid metric handler declared: " + metric.handler);
            return console.trace();
        }
        if (typeof metric.historic !== 'boolean') {
            metric.historic = true;
        }
        this.logger("Registering new metric: " + metric.name);
        this.metrics.set(metric.name, metric);
    };
    MetricService.prototype.meter = function (opts) {
        var metric = {
            name: opts.name,
            type: MetricType.meter,
            id: opts.id,
            historic: opts.historic,
            implementation: new meter_1.default(opts),
            unit: opts.unit,
            handler: function () {
                return this.implementation.isUsed() ? this.implementation.val() : NaN;
            }
        };
        this.registerMetric(metric);
        return metric.implementation;
    };
    MetricService.prototype.counter = function (opts) {
        var metric = {
            name: opts.name,
            type: MetricType.counter,
            id: opts.id,
            historic: opts.historic,
            implementation: new counter_1.default(opts),
            unit: opts.unit,
            handler: function () {
                return this.implementation.isUsed() ? this.implementation.val() : NaN;
            }
        };
        this.registerMetric(metric);
        return metric.implementation;
    };
    MetricService.prototype.histogram = function (opts) {
        if (opts.measurement === undefined || opts.measurement === null) {
            opts.measurement = MetricMeasurements.mean;
        }
        var metric = {
            name: opts.name,
            type: MetricType.histogram,
            id: opts.id,
            historic: opts.historic,
            implementation: new histogram_1.default(opts),
            unit: opts.unit,
            handler: function () {
                return this.implementation.isUsed() ?
                    (Math.round(this.implementation.val() * 100) / 100) : NaN;
            }
        };
        this.registerMetric(metric);
        return metric.implementation;
    };
    MetricService.prototype.metric = function (opts) {
        var metric;
        if (typeof opts.value === 'function') {
            metric = {
                name: opts.name,
                type: MetricType.gauge,
                id: opts.id,
                implementation: undefined,
                historic: opts.historic,
                unit: opts.unit,
                handler: opts.value
            };
        }
        else {
            metric = {
                name: opts.name,
                type: MetricType.gauge,
                id: opts.id,
                historic: opts.historic,
                implementation: new gauge_1.default(),
                unit: opts.unit,
                handler: function () {
                    return this.implementation.isUsed() ? this.implementation.val() : NaN;
                }
            };
        }
        this.registerMetric(metric);
        return metric.implementation;
    };
    MetricService.prototype.deleteMetric = function (name) {
        return this.metrics.delete(name);
    };
    MetricService.prototype.destroy = function () {
        if (this.timer !== null) {
            clearInterval(this.timer);
        }
        this.metrics.clear();
    };
    return MetricService;
}());
exports.MetricService = MetricService;
