// @pm2/io/build/main/services/runtimeStats.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var debug_1 = require("debug");
var module_1 = require("../utils/module");
var eventemitter2_1 = require("eventemitter2");
var RuntimeStatsService = (function (_super) {
    tslib_1.__extends(RuntimeStatsService, _super);
    function RuntimeStatsService() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.logger = debug_1.default('axm:services:runtimeStats');
        _this.enabled = false;
        return _this;
    }
    RuntimeStatsService.prototype.init = function () {
        var _this = this;
        this.logger('init');
        if (process.env.PM2_APM_DISABLE_RUNTIME_STATS === 'true') {
            return this.logger('disabling service because of the environment flag');
        }
        var modulePath = module_1.default.detectModule('@pm2/node-runtime-stats');
        if (typeof modulePath !== 'string')
            return;
        var RuntimeStats = module_1.default.loadModule(modulePath);
        if (RuntimeStats instanceof Error) {
            return this.logger("Failed to require module @pm2/node-runtime-stats: " + RuntimeStats.message);
        }
        this.noduleInstance = new RuntimeStats({
            delay: 1000
        });
        this.logger('starting runtime stats');
        this.noduleInstance.start();
        this.handle = function (data) {
            _this.logger('received runtime stats', data);
            _this.emit('data', data);
        };
        this.noduleInstance.on('sense', this.handle);
        this.enabled = true;
    };
    RuntimeStatsService.prototype.isEnabled = function () {
        return this.enabled;
    };
    RuntimeStatsService.prototype.destroy = function () {
        if (this.noduleInstance !== undefined && this.noduleInstance !== null) {
            this.logger('removing listener on runtime stats service');
            this.noduleInstance.removeListener('sense', this.handle);
            this.noduleInstance.stop();
        }
        this.logger('destroy');
    };
    return RuntimeStatsService;
}(eventemitter2_1.EventEmitter2));
exports.RuntimeStatsService = RuntimeStatsService;
